Add Epsilon- and Eta-Sampling

shermansiu opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Implements epsilon- and eta-sampling, as seen in [[Truncation Sampling as Language Model Desmoothing](https://arxiv.org/abs/2210.15191)](https://arxiv.org/abs/2210.15191). The code is adapted from the author's official repository [here](https://github.com/john-hewitt/truncation-sampling).

Resolves #21092.


## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts and @NielsRogge
- speech models: @sanchit-gandhi

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->
@gante

shermansiu committed:
Add epsilon- and eta-sampling.

Add epsilon- and eta-sampling, following the official code from https://github.com/john-hewitt/truncation-sampling and adapting to be more configurable, as required by Huggingface transformers.

shermansiu committed:
Add unit tests for epsilon- and eta-sampling.

shermansiu committed:
Black: fix code formatting.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21121). All of your documentation changes will be reflected on that endpoint.

shermansiu committed:
Fix docstring spacing.

shermansiu committed:
Clean up newlines.

shermansiu committed:
Fix implementation bugs and their associated tests.

gante commented on src/transformers/generation/logits_process.py:
```diff
@@ -323,6 +320,88 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
+class EpsilonLogitsWarper(LogitsWarper):
+    r"""
+    [`LogitsWarper`] that performs epsilon, i.e. restricting to tokens with absolute prob > prob_cut_off. Takes the
```
```suggestion
    [`LogitsWarper`] that performs epsilon-sampling, i.e. restricting to tokens with absolute prob > prob_cut_off. Takes the
```

(you might need to run `make fixup` locally and commit again, after incorporating this change)

gante commented on src/transformers/generation/logits_process.py:
```diff
@@ -323,6 +320,88 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
+
+class EtaLogitsWarper(LogitsWarper):
+    r"""
+    [`LogitsWarper`] that performs eta-sampling. See [Truncation Sampling as Language Model
```
Some further description of what eta-sampling is would be appreciated here 🙏

shermansiu replied:
Added!

gante commented on src/transformers/generation/configuration_utils.py:
```diff
@@ -109,6 +109,17 @@ class GenerationConfig(PushToHubMixin):
+            Desmoothing](https://arxiv.org/abs/2210.15191) for more details.
+        eta_cutoff (`float`, *optional*, defaults to 0):
+            Eta sampling is a hybrid of locally typical sampling and epsilon sampling. If set to float strictly between
+            0 and 1, a token is only considered if it is greater than either the cutoff or sqrt(eta_cutoff) *
+            expected_next_random_token_probability. In the paper, suggested values range from 3e-4 to 2e-3, depending
```
```suggestion
            0 and 1, a token is only considered if it is greater than either eta_cutoff or sqrt(eta_cutoff) *
            entropy(softmax(next_token_logits)). In the paper, suggested values range from 3e-4 to 2e-3, depending
```

Adding some precision to the terms here. "expected_next_random_token_probability" isn't exactly what the paper uses, if I got it right :) (feel free to correct me if you think I'm wrong!)

shermansiu replied:
It's actually `exp(-entropy(softmax(next_token_logits)))`, but fixed!

gante commented on src/transformers/generation/utils.py:
```diff
@@ -746,23 +748,22 @@ def _get_logits_warper(
         # all samplers can be found in `generation_utils_samplers.py`
         if generation_config.temperature is not None and generation_config.temperature != 1.0:
             warpers.append(TemperatureLogitsWarper(generation_config.temperature))
+        min_tokens_to_keep = 2 if generation_config.num_beams > 1 else 1
```
Sneaky change, but appreciated! 😉

gante reviewed (commented):
Thank you for this great contribution! I've added a few comments, but they are very simple to address. 

My suggested timeline for this PR is:
1. I approve this PR after the suggested changes are implemented
2. Sylvain has a look at this PR and approves it
3. We merge it
4. We build a notebook with a few examples and communicate on twitter/LI. If you are interested, we'd be happy to support you here. If you are not, we can take care of it :)

WDYT?

shermansiu committed:
Remove epsilon- and eta-sampling parameters from PretrainedConfig.

shermansiu committed:
Clarify and clean up the documentation.

shermansiu commented on src/transformers/configuration_utils.py:
```diff
@@ -144,6 +144,17 @@ class PretrainedConfig(PushToHubMixin):
+            0 and 1, a token is only considered if it is greater than either the cutoff or sqrt(eta_cutoff) *
+            expected_next_random_token_probability. In the paper, suggested values range from 3e-4 to 2e-3, depending
+            on the size of the model. See [Truncation Sampling as Language Model
+            Desmoothing](https://arxiv.org/abs/2210.15191) for more details.
```
Gotcha! Removed!

shermansiu commented on src/transformers/configuration_utils.py:
```diff
@@ -286,6 +297,8 @@ def __init__(self, **kwargs):
         self.top_p = kwargs.pop("top_p", 1.0)
         self.typical_p = kwargs.pop("typical_p", 1.0)
+        self.epsilon_cutoff = kwargs.pop("epsilon_cutoff", 0.0)
+        self.eta_cutoff = kwargs.pop("eta_cutoff", 0.0)
```
Done!

shermansiu commented on src/transformers/generation/configuration_utils.py:
```diff
@@ -109,6 +109,17 @@ class GenerationConfig(PushToHubMixin):
         typical_p (`float`, *optional*, defaults to 1.0):
             The amount of probability mass from the original distribution to be considered in typical decoding. If set
             to 1.0 it takes no effect. See [this paper](https://arxiv.org/pdf/2202.00666.pdf) for more details.
+        epsilon_cutoff (`float`, *optional*, defaults to 0):
```
Done!

shermansiu commented on src/transformers/generation/configuration_utils.py:
```diff
@@ -109,6 +109,17 @@ class GenerationConfig(PushToHubMixin):
+            `epsilon_cutoff` will be sampled. In the paper, suggested values range from 3e-4 to 9e-4, depending on the
+            size of the model. See [Truncation Sampling as Language Model
+            Desmoothing](https://arxiv.org/abs/2210.15191) for more details.
+        eta_cutoff (`float`, *optional*, defaults to 0):
```
Done!

shermansiu commented:
@gante I incorporated your changes! I rebased with `main` because there were some merge conflicts, which caused some of the review snippets above to get "outdated," but it's done!

gante reviewed (approved):
LGTM 👍

gante commented:
@sgugger this PR adds a new sampling-based generation strategy that can be effectively implemented through logits processors

shermansiu committed:
Remove parameters for PretrainedConfig test.

sgugger reviewed (approved):
Thanks a lot for your contribution!

gante commented:
@shermansiu now step 4 remains :) Would you like to work on it? (we can retweet and share your posts)

shermansiu commented:
I'm interested, but it feels like it's best added to https://github.com/huggingface/blog/blob/main/notebooks/02_how_to_generate.ipynb than in its own notebook, as other truncation sampling approaches (top-p and top-k) are already there.

gante commented:
We don't update the notebooks for our blog posts, so they stay consistent with their content :) But we will certainly update our [new docs ](https://huggingface.co/docs/transformers/main/en/generation_strategies)

shermansiu commented:
I have a lot on my to-do list right now... Although I'd love to contribute to the notebook, I think it's unrealistic for me to be able to put something out soon.

shermansiu commented:
Yesterday's ACL 2023 tutorial on "Generating Text from Large Language Models" covers eta-sampling and more! John Hewitt, the first author of the eta-sampling paper, was one of the presenters for that tutorial!

Site: https://rycolab.io/classes/acl-2023-tutorial/
Slides: https://drive.google.com/file/d/1UHbGcjzBURG1n2DufC7iDTmGNjIz5Dp_/view
